go/vt/vtgate/vschema_manager.go

```diff
@@ -192,22 +193,21 @@ func (vm *VSchemaManager) updateFromSchema(vschema *vindexes.VSchema) {
 	for ksName, ks := range vschema.Keyspaces {
 		m := vm.schema.Tables(ksName)
 
-		for tblName, columns := range m {
-			vTbl := ks.Tables[tblName]
-			if vTbl == nil {
-				// a table that is unknown by the vschema. we add it as a normal table
-				ks.Tables[tblName] = &vindexes.Table{
-					Name:                    sqlparser.NewIdentifierCS(tblName),
-					Keyspace:                ks.Keyspace,
-					Columns:                 columns,
-					ColumnListAuthoritative: true,
+		for tblName, tblInfo := range m {
+			vTbl := setColumns(ks, tblName, tblInfo.Columns)
+			for _, fkDef := range tblInfo.ForeignKeys {
+				parentTbl, err := vschema.FindRoutedTable(ksName, fkDef.ReferenceDefinition.ReferencedTable.Name.String(), topodatapb.TabletType_PRIMARY)
+				if err != nil {
+					log.Errorf("error finding parent table %s: %v", fkDef.ReferenceDefinition.ReferencedTable.Name.String(), err)
+					continue
 				}
-				continue
-			}
-			if !vTbl.ColumnListAuthoritative {
-				// if we found the matching table and the vschema view of it is not authoritative, then we just update the columns of the table
-				vTbl.Columns = columns
-				vTbl.ColumnListAuthoritative = true
+				childTbl, err := vschema.FindRoutedTable(ksName, tblName, topodatapb.TabletType_PRIMARY)
+				if err != nil {
+					log.Errorf("error finding child table %s: %v", tblName, err)
+					continue
+				}
+				vTbl.ParentForeignKeys = append(vTbl.ParentForeignKeys, vindexes.NewParentFkInfo(parentTbl, fkDef))
+				parentTbl.ChildForeignKeys = append(parentTbl.ChildForeignKeys, vindexes.NewChildFkInfo(childTbl, fkDef))
```

Comment:
Do we not need to check for duplicates?

Reply:
this is coming from MySQL create table statement. We should rely on that not having duplicate constraints.